### ISSUE TITLE
[RFC][Resource] Routing

### ISSUE THREAD

**Issue by aRn0D** [Sylius/Sylius#2559]
Sometime ago, the ApiLoader has been introduced into the resource bundle, it generates the API routes automatically. I think we should use it for the web too because we often write the same routes for simple CRUD.

I update it for a personal project, now it support two kind of the 'routing' (api and web)
``` php  
public function supports($resource, $type = null)
    {
        return in_array($type, ['api', 'web']);
    }
```

it generates these routes (the type is used to generate the key of the route)
``` bash
$ php app/console router:debug | grep theme  
app_web_theme_index            GET       ANY    ANY  /themes/
app_web_theme_show             GET       ANY    ANY  /themes/{id}
app_web_theme_create           POST      ANY    ANY  /themes/
app_web_theme_update           PUT|PATCH ANY    ANY  /themes/{id}
app_web_theme_delete           DELETE    ANY    ANY  /themes/{id}

app_api_theme_index            GET       ANY    ANY  /themes/
app_api_theme_show             GET       ANY    ANY  /themes/{id}
app_api_theme_create           POST      ANY    ANY  /themes/
app_api_theme_update           PUT|PATCH ANY    ANY  /themes/{id}
app_api_theme_delete           DELETE    ANY    ANY  /themes/{id} 
```

If you need to use advance feature (like filtering, sorting, define custom repository etc...) you still can overwrite your route.  

@Sylius/core-team What do you think ?

**Comment by pjedrzejewski** [comment]
Definitely in favor! I'd vote for having api and "generic" loading separated and make the "prefix" (in name of route, not url prefix) configurable, so that we can generate routing for Sylius backend as well.

**Comment by aRn0D** [comment]
You spoke about creating two classes : ApiLoader and a GenericLoader? We can inject the prefix in the constructor easily, we should keep the pattern `appName_prefix_resourceName_index`.

**Comment by pjedrzejewski** [comment]
Yeah, maybe 2 loaders or one, but just more flexible. Not sure. I think we can pass array configuration to ``resource`` when loading, no? That would allow us for even more configurations.

**Comment by aRn0D** [comment]
I think it is possible to manage it with one loader! What do you means by 'pass array configuration to resource'? inject `sylius.config.classes` into the loader?

**Comment by aRn0D** [comment]
Version 2 : 
Create two route loader : ApiLoader (already exists) and GenericLoader (used for the web)

ApiLoader
Type : sylius.api

Generated routes :
``` bash
$ php app/console router:debug | grep resource  
app_api_resource_index            GET       ANY    ANY  /resources/
app_api_resource_show             GET       ANY    ANY  /resources/{id}
app_api_resource_create           POST      ANY    ANY  /resources/
app_api_resource_update           PUT|PATCH ANY    ANY  /resources/{id}
app_api_resource_delete           DELETE    ANY    ANY  /resources/{id} 
```
`resources` is the plurialized name of the resource


GenericLoader
Type : sylius

``` bash
$ php app/console router:debug | grep resource  
app_resource_index                GET       ANY    ANY  /resources/        
app_resource_show                GET       ANY    ANY  /resources/{id}    
app_resource_delete               DELETE    ANY    ANY  /resources/{id}    
app_resource_create               GET|POST  ANY    ANY  /resources/new     
app_resource_update               GET|POST  ANY    ANY  resources/{id}/edit
```
`resources` is the plurialized name of the resource

`sylius.resource.classes` should be injected in their constructors, they should get all configured resources and build their routes.

Idea:  A `RouteFactory` should be created for building easily a `Route`

**Comment by aRn0D** [comment]
@Sylius/core-team What do you think about the version 2 ?

**Comment by aRn0D** [comment]
**Version 3 :**
Create two route loaders : ApiLoader (already exists but it should be refactored) and GenericLoader (used for the web)

## ApiLoader
**Type :** sylius.api

Generated routes :
``` bash
$ php app/console router:debug | grep resource  
app_api_resource_index            GET       ANY    ANY  /resources/
app_api_resource_show             GET       ANY    ANY  /resources/{id}
app_api_resource_create           POST      ANY    ANY  /resources/
app_api_resource_update           PUT|PATCH ANY    ANY  /resources/{id}
app_api_resource_delete           DELETE    ANY    ANY  /resources/{id} 
```
`resources` is the pluralized name of the resource
Route name pattern : `(appName)_api_(resourceName)_(actionName)`

## GenericLoader
**Type :** sylius.route_prefix, the route prefix can be empty and it will be used to build the key of the route.

``` bash
$ php app/console router:debug | grep resource  
app_route_prefix_resource_index                GET       ANY    ANY  /resources/        
app_route_prefix_resource_show                GET       ANY    ANY  /resources/{id}    
app_route_prefix_resource_delete               DELETE    ANY    ANY  /resources/{id}    
app_route_prefix_resource_create               GET|POST  ANY    ANY  /resources/new     
app_route_prefix_resource_update               GET|POST  ANY    ANY  resources/{id}/edit
```
`resources` is the pluralized name of the resource
Route name pattern : `(appName)_(resourceName)_(actionName)`

## RouteFactory
A `RouteFactory` should be created to build easily a `Route`, it should avoid code duplication.

## Constructor
The `RouteFactory` and `sylius.resource.classes` (container parameter) should be injected in their constructors. The second parameter should be parsed to get all configured resources and build their routes.

## Activation
```yaml
api_block:
    resource: .
    type: sylius
```
or
```yaml
api_block:
    resource: .
    type: sylius.backend
```
or
```yaml
api_block:
    resource: .
    type: sylius.api
```

@Sylius/core-team : any opinion?

**Comment by pjedrzejewski** [comment]
Can we pass an array of options to ``resource`` instead of ``.``. That would allow us for more configurations. And keep the type for sylius generic loader just ``sylius``.

**Comment by aRn0D** [comment]
**Version 4 :**
Create two route loaders : ApiLoader (already exists but it should be refactored) and GenericLoader (used for the web)

## ApiLoader
**Type :** sylius.api

Generated routes :
``` bash
$ php app/console router:debug | grep resource  
app_api_resource_index            GET       ANY    ANY  /resources/
app_api_resource_show             GET       ANY    ANY  /resources/{id}
app_api_resource_create           POST      ANY    ANY  /resources/
app_api_resource_update           PUT|PATCH ANY    ANY  /resources/{id}
app_api_resource_delete           DELETE    ANY    ANY  /resources/{id} 
```
`resources` is the pluralized name of the resource
Route name pattern : `(appName)_api_(resourceName)_(actionName)`

```yaml
api_block:
    resource: .
    type: sylius.api
```

## GenericLoader
**Type :** sylius
**Resource :** an array of option : 
* prefix: used to build the key of the route (optional)
 
``` bash
$ php app/console router:debug | grep resource  
app_route_prefix_resource_index                GET       ANY    ANY  /resources/        
app_route_prefix_resource_show                GET       ANY    ANY  /resources/{id}    
app_route_prefix_resource_delete               DELETE    ANY    ANY  /resources/{id}    
app_route_prefix_resource_create               GET|POST  ANY    ANY  /resources/new     
app_route_prefix_resource_update               GET|POST  ANY    ANY  resources/{id}/edit
```
`resources` is the pluralized name of the resource
Route name pattern : `(appName)_(prefix)_(resourceName)_(actionName)`

```yaml
api_block:
    resource: []
    type: sylius
```

## RouteFactory
A `RouteFactory` should be created to build easily a `Route`, it should avoid code duplication.

## Constructor
The `RouteFactory` and `sylius.resource.classes` (container parameter) should be injected in their constructors. The second parameter should be parsed to get all configured resources and build their routes.


@Sylius/core-team : any opinion?

**Comment by aRn0D** [comment]
@Sylius/core-team / @stloyd What do you think about it ?

**Comment by stloyd** [comment]
Last proposal sounds reasonable, I would like to see the PoC for future discussion about details.

**Comment by aRn0D** [comment]
I am on it !

**Comment by michalmarcinkowski** [comment]
Implemented in #3776
